src/vivarium_census_prl_synth_pop/results_processing/ssn_and_itin.py

```diff
@@ -24,46 +67,166 @@ def generate_itin(
+        itin["serial"] = random_integers(
+            min_val=1,
+            max_val=9999,
+            index=itin.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_itin_serial",
+        )
+        itin["itin"] = ""
+        itin["itin"] += (
+            itin["area"].astype(str) + "-"
+        )  # no zfill needed, values should be 900-999
+        itin["itin"] += itin["group"].astype(str) + "-"  # no zfill needed likewise
+        itin["itin"] += itin["serial"].astype(str).str.zfill(4)
+
+        return itin["itin"].to_numpy()
+
+    if additional_key is None:
+        additional_key = 1
+    itins = pd.Series(generate_itin(size, additional_key))
+    duplicate_mask = itins.duplicated()
+    counter = 0
+    while duplicate_mask.sum() > 0:
+        if counter >= 10:
+            logger.info(
+                f"Resampled ITIN {counter} times and data still contains {duplicate_mask.sum()}"
+                f"duplicates remaining.  Check data or usage of randomness stream."
+            )
+            break
+        new_additional_key = f"{additional_key}_{counter}"
+        itins.loc[duplicate_mask] = generate_itin(duplicate_mask.sum(), new_additional_key)
+        duplicate_mask = itins.duplicated()
+        counter += 1
+    return itins.to_numpy()
+
+
+def generate_ssns(
+    size: int,
+    additional_key: Any,
+    randomness: RandomnessStream,
+) -> np.ndarray:
+    """
+    Generate a np.ndarray of length `size` of unique SSN values.
+
+        first three digits 1 and 899 (inclusive), 666 disallowed, followed by dash
+        next two digits between 1-99 (inclusive), followed by a dash
+        last four digits between 1-9999 (inclusive)
+
+    :param size: The number of itins to generate
+    :param additional_key: Additional key to be used by randomness
+    :param randomness: RandomnessStream stream to use
+
+    """
+
+    def generate_ssn(
+        count: int,
+        extra_additional_key: Any,
+    ) -> pd.DataFrame:
+        ssn = pd.DataFrame(index=pd.Index(range(count)))
 
-    # Rescale the group digits to match expected ranges
-    pre_shift_groups = itin["group"].copy()
-    itin.loc[pre_shift_groups <= 16, "group"] += 49
-    itin.loc[((pre_shift_groups <= 35) & (pre_shift_groups >= 17)), "group"] += 53
-    itin.loc[((pre_shift_groups <= 38) & (pre_shift_groups >= 36)), "group"] += 54
-    itin.loc[pre_shift_groups >= 39, "group"] += 55
-
-    # Serial numbers 1-9999 inclusive
-    itin["serial"] = random_integers(
-        min_val=1,
-        max_val=9999,
-        index=itin.index,
-        randomness=randomness,
-        additional_key=f"{additional_key}_itin_serial",
+        area = random_integers(
+            min_val=1,
+            max_val=899,
+            index=ssn.index,
+            randomness=randomness,
+            additional_key=f"{extra_additional_key}_ssn_area",
+        )
+        area = np.where(area == 666, 667, area)  # XXX is this reasonable?
```

Comment:
why no 666?

Reply:
Oh, nm - they just don't assign that prefix. Yeah, seems reasonable unless `area.replace(666, 667)` is more performant

Reply:
I forgot I left this comment! I think I had that comment as a question for RT, which I've got answered.